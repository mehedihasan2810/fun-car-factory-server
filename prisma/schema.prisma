generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model User {
  id    String  @id @default(auto()) @map("_id") @db.ObjectId
  name  String?
  email String  @unique
  role  String  @default("user")

  @@map("users")
}

model Car {
  id          String @id @default(auto()) @map("_id") @db.ObjectId
  category    String
  description String
  email       String
  name        String
  price       Int
  quantity    Int
  rating      Int
  sellerName  String
  url         String

  @@map("cars")
}
