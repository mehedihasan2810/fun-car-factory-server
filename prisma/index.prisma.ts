import { PrismaClient } from "@prisma/client";

export const prisma = new PrismaClient();

// log the time that our queries are taking start
// prisma.$use(async (params, next) => {
//   const before = Date.now();

//   const result = await next(params);

//   const after = Date.now();

//   console.log(
//     `Query ${params.model}.${params.action} took ${after - before}ms`
//   );

//   return result;
// });
// log the time that our queries are taking end
