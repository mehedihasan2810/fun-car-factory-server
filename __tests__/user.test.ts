import { testServer } from ".";
import { prisma } from "../prisma/index.prisma";

const GET_USERS = `#graphql 
query GetUsers {
     getUsers {
        name
        email
        role
     } 
    }`;

const GET_USER = `#graphql 
query GetUser($email: String!) {
     getUser(email: $email) {
        name
        email
        role
     } 
    }`;

const CREATE_USER = `#graphql
mutation CreateUser($input: UserInput!) {
     createUser(input: $input) {
        email
        name
        role
     }
    }`;

const mockUser = {
  email: "usertestjest@gmail.com",
  name: "mehedi",
  role: "user",
};

describe("User Queries", () => {
  beforeAll(async () => {
    const isUserExists = await prisma.user.findUnique({
      where: {
        email: mockUser.email,
      },
    });

    if (isUserExists) {
      await prisma.user.delete({
        where: {
          email: mockUser.email,
        },
      });
    }
  }, 120000);

  test("Should fetch a list of users", async () => {
    const getUsersRes = await testServer.executeOperation({
      query: GET_USERS,
    });

    expect((getUsersRes.body as any).singleResult.errors).toBeUndefined();
    expect(getUsersRes).toMatchSnapshot();
  }, 120000);

  test("Should fetch a single user based on a unique email", async () => {
    const getUserRes = await testServer.executeOperation({
      query: GET_USER,
      variables: { email: "test@gmail.com" },
    });

    expect((getUserRes.body as any).singleResult.errors).toBeUndefined();
    expect(getUserRes).toMatchSnapshot();
  }, 120000);
});

describe("User Mutations", () => {
  afterAll(async () => {
    await prisma.user.delete({
      where: {
        email: mockUser.email,
      },
    });
  }, 120000);

  test("should create a user with the field email, name, role and return the created data", async () => {
    const createUserRes = await testServer.executeOperation({
      query: CREATE_USER,
      variables: {
        input: mockUser,
      },
    });

    const user = await prisma.user.findUnique({
      where: {
        email: mockUser.email,
      },
      select: {
        email: true,
        name: true,
        role: true,
      },
    });
    expect(user).toEqual(mockUser);
    expect((createUserRes.body as any).singleResult.errors).toBeUndefined();
    // expect(createUserRes).toMatchSnapshot();
  }, 120000);
});
