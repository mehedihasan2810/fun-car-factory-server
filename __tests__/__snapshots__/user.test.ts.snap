// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Queries Should fetch a list of users 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "getUsers": [
          {
            "email": "test@gmail.com",
            "name": "Micheal Clark",
            "role": "admin",
          },
          {
            "email": "test2@gmail.com",
            "name": "Stuart Board",
            "role": "user",
          },
          {
            "email": "shakib@gmail.com",
            "name": "shakib",
            "role": "user",
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User Queries Should fetch a single user based on a unique email 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "getUser": {
          "email": "test@gmail.com",
          "name": "Micheal Clark",
          "role": "admin",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
